### ISSUE TITLE
[Question] Transforms for NestedTensor

### ISSUE THREAD

**Issue by voldemortX** [facebookresearch/detr#373]
@alcinos  Hi! I'm investigating the use of NestedTensor to represent image and padding mask for transformers.
As I currently understood, padding masks should start with all 0 and padded with 1. While the image will be padded with 0. But I can't find any special treatment for this datatype in [transforms](https://github.com/facebookresearch/detr/blob/master/datasets/transforms.py). Also I'd like to understand more about how are NestedTensors collated when loading data? 

 I kind of needs to do rotations in my own code, so I'm not sure how this should be done properly. Currently, I don't use NestedTensors and put padding mask together with all the targets in a dict, and keep the dict in batching (i.e. batched as `list[dict]`). But that seems to be inefficient and it tends to run out my system file descriptors when multi-processing the `DataLoader` with large `num_workers`.

**Comment by alcinos** [comment]
The NestedTensors are created when collating, hence after all the transforms are done.
The masking is used exclusively for the padding that is added during the collate, that is if the images in the batch have different shapes.
If your transformation involves adding padding (let's call it transform-padding to avoid confusion), then you should NOT mask it out. You should make sure, however, that you do modify the image size info in the target dictionary so that it reflects any increase of size due to this transform-padding. 

To summarize, if your rotations add some transform-padding (presumably because you rotate in angles that are not 90/180/270), you do not need to worry about it, just make sure that the size info is correct.

Hope this helps
Best of luck

**Comment by voldemortX** [comment]
Thanks very much! Your answers really helped me a lot.

> 
> To summarize, if your rotations add some transform-padding (presumably because you rotate in angles that are not 90/180/270), you do not need to worry about it, just make sure that the size info is correct.
> 

Just to be sure, if my application does not concern multi-scale training/testing, then I don't need padding mask at all? I guess the rational here would be: no one masked out anything for transform-padding when using CNN.

**Comment by alcinos** [comment]
> Just to be sure, if my application does not concern multi-scale training/testing, then I don't need padding mask at all? I guess the rational here would be: no one masked out anything for transform-padding when using CNN.

You don't need to worry about padding mask if and only if all the images that comes out of the set of transform that you have all have the exact same size. So if all your images are initially the same sizes and you remove the multi-scale/crop augmentations that we have, then it's fine.
I would strongly advise against removing multi-scale/crops though. They are quite important to avoid overfitting and improve the performance quite a bit.

It's true that the CNN will always "see" the padding as well as the transform-padding (no distinction between the two). However, the boxes location and size are normalized with respect to the image size *before* collating (that is, including transform-padding but not collate padding). Because of that, the transformer needs to "know" what was the original image size, and to do that, we simply mask out what isn't part of the original image.
Hope this helps, feel free to let me know if you have further questions.

**Comment by voldemortX** [comment]
Thanks for the detailed answers! Now I understand this perfectly.
